cns/restserver/ipam.go

```diff
@@ -478,101 +634,194 @@ func (service *HTTPRestService) MarkExistingIPsAsPendingRelease(pendingIPIDs []s
-			}
+		_, found := ncMap[ipState.NCID]
+		// Checks if we haven't already found an IP from that NC
+		if found {
+			continue
+		}
+		// Checks if the current IP is available
+		if ipState.GetState() != types.Available {
+			continue
+		}
+		ncMap[ipState.NCID] = ipState
+		// Once one IP per container is found break out of the loop and stop searching
+		if len(ncMap) == len(service.state.ContainerStatus) {
+			break
+		}
+	}
 
-			podIPInfo := cns.PodIpInfo{}
-			if err := service.populateIPConfigInfoUntransacted(ipState, &podIPInfo); err != nil {
-				return cns.PodIpInfo{}, err
-			}
+	podIPctr := 0
+	// assigns all IPs in the map to the pod
+	for id := range ncMap {
+
+		if err := service.assignIPConfig(ncMap[id], podInfo); err != nil {
+			logger.Errorf(err.Error())
+			break
+		}
 
+		if err := service.populateIPConfigInfoUntransacted(ncMap[id], &podIPInfo[podIPctr]); err != nil {
+			logger.Errorf(err.Error())
+			break
+		}
+		podIPctr++
+		// Checks if we no longer have any IPs to be added and returns if we don't
+		if podIPctr == len(service.state.ContainerStatus) {
 			return podIPInfo, nil
 		}
 	}
+
+	// if we were able to find at least one IP but not enough
+	if len(ncMap) > 0 {
+		logger.Printf("[AssignAvailableIPConfigs] Failed to retrieve enough IPs. Releasing all IPs that were found")
+		for _, ipState := range ncMap { //nolint:gocritic // ignore copy
+			_, err := service.unassignIPConfig(ipState, podInfo)
+			if err != nil {
+				return podIPInfo, fmt.Errorf("[AssignAvailableIPConfigs] failed to mark IPConfig [%+v] back to Available. err: %w", ipState, err)
+			}
+		}
+	}
 	//nolint:goerr113
-	return cns.PodIpInfo{}, fmt.Errorf("no IPs available, waiting on Azure CNS to allocate more")
+	return podIPInfo, fmt.Errorf("not enough IPs available, waiting on Azure CNS to allocate more")
 }
 
 // If IPConfig is already assigned to pod, it returns that else it returns one of the available ipconfigs.
-func requestIPConfigHelper(service *HTTPRestService, req cns.IPConfigRequest) (cns.PodIpInfo, error) {
+func requestIPConfigHelper(service *HTTPRestService, req cns.IPConfigsRequest) ([]cns.PodIpInfo, error) {
 	// check if ipconfig already assigned tothis pod and return if exists or error
 	// if error, ipstate is nil, if exists, ipstate is not nil and error is nil
-	podInfo, err := cns.NewPodInfoFromIPConfigRequest(req)
+	podInfo, err := cns.NewPodInfoFromIPConfigsRequest(req)
 	if err != nil {
-		return cns.PodIpInfo{}, errors.Wrapf(err, "failed to parse IPConfigRequest %v", req)
+		return []cns.PodIpInfo{}, errors.Wrapf(err, "failed to parse IPConfigsRequest %v", req)
 	}
 
 	if podIPInfo, isExist, err := service.GetExistingIPConfig(podInfo); err != nil || isExist {
 		return podIPInfo, err
 	}
 
 	// return desired IPConfig
-	if req.DesiredIPAddress != "" {
-		return service.AssignDesiredIPConfig(podInfo, req.DesiredIPAddress)
+	if len(req.DesiredIPAddresses) != 0 {
+		if req.DesiredIPAddresses[0] != "" {
```

Comment:
why only check 0th index?
What if the len of that slice is 2 but second desiredIpAddress value is empty?

Reply:
revised to include a `validateDesiredIPAddresses` function that checks all indices.